linkerd/reconnect/src/lib.rs

```diff
@@ -3,34 +3,190 @@
 }
 
-impl<C, R, T> NewService<T> for NewReconnect<C, R>
+#[derive(Debug)]
+enum State<B, S> {
+    Disconnected {
+        backoff: Option<B>,
+    },
+    Pending {
+        service: Option<S>,
+        backoff: Option<B>,
+    },
+    Connected(S),
+}
+
+// === impl NewReconnect ===
+
+impl<R: Clone, N> NewReconnect<R, N> {
+    pub fn layer(recover: R) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| NewReconnect {
+            inner,
+            recover: recover.clone(),
+        })
+    }
+}
+
+impl<T, R, N> NewService<T> for NewReconnect<R, N>
 where
     R: Recover + Clone,
-    C: tower::Service<T> + Clone,
+    N: NewService<T> + Clone,
 {
-    type Service = service::Service<T, R, C>;
+    type Service = Reconnect<T, R, N>;
 
     fn new_service(&mut self, target: T) -> Self::Service {
-        service::Service::new(target, self.connect.clone(), self.recover.clone())
+        Reconnect::new(target, self.inner.clone(), self.recover.clone())
+    }
+}
+
+// === impl Reconnect ===
+
+impl<T, R, N> Reconnect<T, R, N>
+where
+    R: Recover,
+    N: NewService<T>,
+{
+    pub fn new(target: T, inner: N, recover: R) -> Self {
+        Self {
+            target,
+            inner,
+            recover,
+            state: State::Disconnected { backoff: None },
+        }
+    }
+}
+
+impl<T, Req, R, N, S> Service<Req> for Reconnect<T, R, N>
+where
+    T: Clone,
+    R: Recover,
+    R::Backoff: Unpin,
+    N: NewService<T, Service = S>,
+    S: Service<Req>,
+    S::Error: Into<Error>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = future::MapErr<S::Future, fn(S::Error) -> Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        loop {
+            self.state = match self.state {
+                // If the service is connected, ensure it's ready. If checking the readiness fails,
+                // go into the disconnected state to poll the backoff before creating a new service.
+                State::Connected(ref mut svc) => match ready!(svc.poll_ready(cx)) {
+                    Ok(()) => {
+                        trace!("Ready");
+                        return Poll::Ready(Ok(()));
+                    }
+                    Err(e) => {
+                        // If the service fails, try to recover.
+                        let error: Error = e.into();
+                        warn!(%error, "Service failed");
+                        let backoff = self.recover.recover(error)?;
+                        debug!("Recovering");
+                        State::Disconnected {
+                            backoff: Some(backoff),
+                        }
+                    }
+                },
+
+                // If the service has not yet become ready, check its readiness and, if that fails,
+                // retain the prior backoff. Otherwise, mark the service as connected. We'll avoid
```

Comment:
it looks like this is a behavioral change from the previous version, which didn't reset the backoff when `poll_ready` errored? but, this seems right, since the service itself is probably a  `Pending`?

Reply:
Previously only `MakeService::poll_ready` wasn't handled with recovery. But now that has been eliminated by using `NewService`. In practice, that will be folded into service's `poll_ready` and handled with recovery. I don't think this matters for us.